### PR TITLE
Update to OCaml 4.12

### PR DESCRIPTION
**Pull request by mnxn**
CI was failing because the master branch of ocaml-lsp requires 4.12.

Also pins gen_js_api to 1.06 because 1.0.7 introduces some breaking changes.

### CODE REVIEW

**Comment by mnxn** [comment]
```
The following dependencies couldn't be met:
  - vscode → gen_js_api = 1.0.6 → ocaml-migrate-parsetree < 2.0.0 → ocaml < 4.12
      base of this switch (use `--unlock-base' to force)
  - ocamlformat → base < v0.12 → ocaml < 4.08.0
      base of this switch (use `--unlock-base' to force)
  - ocamlformat → ocaml < 4.12
      base of this switch (use `--unlock-base' to force)
  - ocamlformat → ocaml-migrate-parsetree >= 2.1.0
  - ocamlformat → ocamlformat_support < 0.5 → ocaml < 4.08.0
      base of this switch (use `--unlock-base' to force)
 ```
 
 Darn. It seems more complicated than just updating the OCaml version.

**Comment by mnxn** [comment]
Upgrading gen_js_api to 1.0.7 is blocked by https://github.com/LexiFi/gen_js_api/issues/148

**Comment by smorimoto** [comment]
I think esy related stuff needs to be updated as well.

File: esy.json [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
+    "@opam/js_of_ocaml": "3.8.0",
     "@opam/gen_js_api": "1.0.6",
-    "@opam/core_kernel": "v0.14.0",
+    "@opam/core_kernel": "v0.14.1",
     "@opam/promise_jsoo": "0.3.1",
     "@opam/jsonoo": "0.2.1",
     "@opam/opam-file-format": "2.1.2"
   },
   "devDependencies": {
     "@opam/dune": "2.7.1",
```
Comment (smorimoto) [inline esy.json:12]:
```suggestion
    "@opam/dune": "*",
```

Comment (smorimoto) [reply esy.json]: IIRC, we can ignore this, as this was probably pinned by a limitation when we were using OCaml 4.06.x.

**Review by smorimoto — APPROVED** [review]

**Pull request merged**